File: tools/nebula-spark/src/main/java/com/vesoft/nebula/reader/NebulaRelation.java
/* Copyright (c) 2020 vesoft inc. All rights reserved.
 *
 * This source code is licensed under Apache 2.0 License,
 * attached with Common Clause Condition 1.0, found in the LICENSES directory.
 */

package com.vesoft.nebula.reader;

import com.facebook.thrift.TException;
import com.vesoft.nebula.bean.DataSourceConfig;
import com.vesoft.nebula.client.meta.MetaClientImpl;
import com.vesoft.nebula.common.Type;

import java.io.Serializable;
import java.util.List;
import java.util.Map;
import java.util.HashMap;
import java.util.ArrayList;

import com.vesoft.nebula.exception.GraphConnectException;
import com.vesoft.nebula.util.DataTypeConverter;
import org.apache.spark.rdd.RDD;
import org.apache.spark.sql.Row;
import org.apache.spark.sql.SQLContext;
import org.apache.spark.sql.sources.BaseRelation;
import org.apache.spark.sql.sources.TableScan;
import org.apache.spark.sql.types.DataTypes;
import org.apache.spark.sql.types.StructField;
import org.apache.spark.sql.types.StructType;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

public class NebulaRelation extends BaseRelation implements Serializable, TableScan {

    private static final Logger LOGGER = LoggerFactory.getLogger(NebulaRelation.class);

    private SQLContext sqlContext;
    private StructType schema;
    private final Map<String, List<StructField>> labelFields = new HashMap<>();

    private DataSourceConfig dataSourceConfig;

    public NebulaRelation(SQLContext sqlContext, DataSourceConfig dataSourceConfig) {
        this.sqlContext = sqlContext;
        this.dataSourceConfig = dataSourceConfig;

        initSchema(dataSourceConfig);
    }

    /**
     * init result dataset's schema
     *
     * @param dataSourceConfig
     */
    private void initSchema(DataSourceConfig dataSourceConfig) {
        Map<String, List<String>> returnColMap = dataSourceConfig.getReturnColMap();
        LOGGER.info("return col map: {}", returnColMap);

        List<StructField> fields = new ArrayList<>();

        MetaClientImpl metaClient = new MetaClientImpl(dataSourceConfig.getHostAndPorts());
        try {
            metaClient.connect();
        } catch (TException e) {
            throw new GraphConnectException(e.getMessage(), e);
        }

        Map<String, Class> schemaColAndType;
        for (Map.Entry<String, List<String>> returnColEntry : returnColMap.entrySet()) {
            if (Type.VERTEX.getType().equalsIgnoreCase(dataSourceConfig.getType())) {
                fields.add(DataTypes.createStructField("_vertexId", DataTypes.StringType, false));
                schemaColAndType = metaClient.getTagSchema(dataSourceConfig.getNameSpace(), dataSourceConfig.getLabel());
            } else {
                fields.add(DataTypes.createStructField("_srcId", DataTypes.StringType, false));
                fields.add(DataTypes.createStructField("_dstId", DataTypes.StringType, false));
                schemaColAndType = metaClient.getEdgeSchema(dataSourceConfig.getNameSpace(), dataSourceConfig.getLabel());
            }

            if (dataSourceConfig.getAllCols()) {
                // if allCols is true, then fields should contain all properties.
                for (Map.Entry<String, Class> colTypeEntry : schemaColAndType.entrySet()) {
                    fields.add(DataTypes.createStructField(colTypeEntry.getKey(),
                            DataTypeConverter.convertDataType(colTypeEntry.getValue()),
                            true));
                }
            } else {
                for (String returnCol : returnColEntry.getValue()) {
                    if (schemaColAndType.containsKey(returnCol)) {
                        fields.add(DataTypes.createStructField(returnCol,
                                DataTypeConverter.convertDataType(schemaColAndType.get(returnCol)),
                                true));
                    } else {
                        LOGGER.warn("label {} doesn't contain col {}", dataSourceConfig.getLabel(), returnCol);
                    }
                }
            }

            labelFields.put(returnColEntry.getKey(), fields);
            schema = new StructType(fields.toArray(new StructField[fields.size()]));
            LOGGER.info("return prop set for label {} : {}", returnColEntry.getKey(), fields);
        }
    }

    @Override
    public SQLContext sqlContext() {
        return sqlContext;
    }

    @Override
    public StructType schema() {
        return schema;
    }

    @Override
    public RDD<Row> buildScan() {
        return new NebulaRDD(sqlContext, dataSourceConfig);
    }
}

File: tools/nebula-spark/src/main/scala/com/vesoft/nebula/tools/connector/reader/NebulaRelation.scala
/* Copyright (c) 2020 vesoft inc. All rights reserved.
 *
 * This source code is licensed under Apache 2.0 License,
 * attached with Common Clause Condition 1.0, found in the LICENSES directory.
 */

package com.vesoft.nebula.tools.connector.reader

import com.vesoft.nebula.tools.connector.{DataTypeEnum, NebulaOptions, NebulaUtils}
import com.vesoft.nebula.ColumnDef
import com.vesoft.nebula.utils.NebulaTypeUtil
import org.apache.spark.rdd.RDD
import org.apache.spark.sql.sources.{BaseRelation, TableScan}
import org.apache.spark.sql.types.{DataType, DataTypes, StructField, StructType}
import org.apache.spark.sql.{Row, SQLContext}
import org.slf4j.LoggerFactory

import scala.collection.mutable
import scala.collection.mutable.ListBuffer

private case class NebulaRelation(override val sqlContext: SQLContext, nebulaOptions: NebulaOptions)
    extends BaseRelation
    with TableScan {
  private val LOG = LoggerFactory.getLogger(this.getClass)

  private val labelFields: Map[String, List[StructField]] = Map()

  private var datasetSchema: StructType = _

  override val needConversion: Boolean = false

  override def schema: StructType = getSchema(nebulaOptions)

  /**
    * convert [[ColumnDef]] to spark [[DataType]]
    */
  private def columnDef2dataType(columnDef: ColumnDef): DataType = {
    NebulaUtils.convertDataType(NebulaTypeUtil.supportedTypeToClass(columnDef.getType.getType))
  }

  /**
    * return the dataset's schema. Schema includes configured cols in returnCols or includes all properties in nebula.
    */
  def getSchema(nebulaOptions: NebulaOptions): StructType = {
    val returnColMap                    = nebulaOptions.getReturnColMap
    val fields: ListBuffer[StructField] = new ListBuffer[StructField]
    val metaClient                      = NebulaUtils.createMetaClient(nebulaOptions.getHostAndPorts, nebulaOptions)

    import scala.collection.JavaConverters._
    var schemaCols: Seq[ColumnDef] = Seq()

    returnColMap.keySet.foreach(k => {
      val schema = if (DataTypeEnum.VERTEX.toString.equalsIgnoreCase(nebulaOptions.dataType)) {
        fields.append(DataTypes.createStructField("_vertexId", DataTypes.StringType, false))
        metaClient.getTag(nebulaOptions.spaceName, nebulaOptions.label)
      } else {
        fields.append(DataTypes.createStructField("_srcId", DataTypes.StringType, false))
        fields.append(DataTypes.createStructField("_dstId", DataTypes.StringType, false))
        metaClient.getEdge(nebulaOptions.spaceName, nebulaOptions.label)
      }
      schemaCols = schema.columns.asScala
      if (nebulaOptions.allCols) {
        // if allCols is true, then fields should contain all properties.
        schemaCols
          .foreach(columnDef => {
            LOG.info(s"prop name ${columnDef.getName}, type ${columnDef.getType}")
            fields.append(DataTypes.createStructField(columnDef.getName, columnDef2dataType(columnDef), true))
          })
      } else {
        // create map, columnName -> columnDef
        val colName2Def = schemaCols.groupBy(_.getName).mapValues(_.last)
        returnColMap(k)
          .foreach(returnCol => {
            colName2Def.get(returnCol) match {
              case Some(columnDef) =>
                fields.append(DataTypes.createStructField(returnCol, columnDef2dataType(columnDef), true))
              case None => LOG.warn(s"label ${nebulaOptions.label} doesn't contain col $returnCol")
            }
          })
      }
      labelFields ++ Map(k -> fields)
      datasetSchema = new StructType(fields.toArray)
    })
    LOG.info(s"dataset's schema: $datasetSchema")
    datasetSchema
  }

  override def buildScan(): RDD[Row] = {
    new NebulaRDD(sqlContext, nebulaOptions, datasetSchema).asInstanceOf[RDD[Row]]
  }
}
